x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/EncryptedRepository.java

```diff
@@ -6,11 +6,325 @@
+    static final String DATA_ENCRYPTION_SCHEME = "AES/GCM/NoPadding";
+    static final int DATA_KEY_SIZE_IN_BITS = 256;
     static final long PACKET_START_COUNTER = Long.MIN_VALUE;
-    static final int MAX_PACKET_LENGTH_IN_BYTES = 1 << 30;
+    static final int MAX_PACKET_LENGTH_IN_BYTES = 1 << 20; // 1MB
+    static final int PACKET_LENGTH_IN_BYTES = 64 * (1 << 10); // 64KB
+
+    private static final String ENCRYPTION_METADATA_PREFIX = "encryption-metadata";
+
+    private final BlobStoreRepository delegatedRepository;
+    private final KeyGenerator dataEncryptionKeyGenerator;
+    private final PasswordBasedEncryptor metadataEncryptor;
+    private final ConsistentSettingsService consistentSettingsService;
+    private final SecureRandom secureRandom;
+    private final SecureSetting<?> passwordSettingForThisRepo;
+
+    protected EncryptedRepository(RepositoryMetaData metadata, NamedXContentRegistry namedXContentRegistry, ClusterService clusterService,
+                                  BlobStoreRepository delegatedRepository, PasswordBasedEncryptor metadataEncryptor,
+                                  ConsistentSettingsService consistentSettingsService) throws NoSuchAlgorithmException {
+        super(metadata, namedXContentRegistry, clusterService, delegatedRepository.basePath());
+        this.delegatedRepository = delegatedRepository;
+        this.dataEncryptionKeyGenerator = KeyGenerator.getInstance(EncryptedRepositoryPlugin.CIPHER_ALGO);
+        this.dataEncryptionKeyGenerator.init(DATA_KEY_SIZE_IN_BITS, SecureRandom.getInstance(EncryptedRepositoryPlugin.RAND_ALGO));
+        this.metadataEncryptor = metadataEncryptor;
+        this.consistentSettingsService = consistentSettingsService;
+        this.secureRandom = SecureRandom.getInstance(EncryptedRepositoryPlugin.RAND_ALGO);
+        this.passwordSettingForThisRepo =
+                (SecureSetting<?>) EncryptedRepositoryPlugin.ENCRYPTION_PASSWORD_SETTING.getConcreteSettingForNamespace(metadata.name());
+    }
+
+    @Override
+    public void snapshotShard(Store store, MapperService mapperService, SnapshotId snapshotId, IndexId indexId,
+                              IndexCommit snapshotIndexCommit, IndexShardSnapshotStatus snapshotStatus, boolean writeShardGens,
+                              ActionListener<String> listener) {
+        if (EncryptedRepositoryPlugin.getLicenseState().isEncryptedRepositoryAllowed()) {
+            if (consistentSettingsService.isConsistent(passwordSettingForThisRepo)) {
+                super.snapshotShard(store, mapperService, snapshotId, indexId, snapshotIndexCommit, snapshotStatus, writeShardGens, listener);
+            } else {
+                listener.onFailure(new RepositoryException(metadata.name(),
+                        "The local node's value for the keystore secure setting [" + passwordSettingForThisRepo.getKey() + "] does not " +
+                                "match the master's"));
+            }
```

Comment:
Hi @tvernum @jkakavas @original-brownbear ! I feel I need your input regarding the password consistency check when snapshotting.

This check here assures that a snapshot cannot contain shards which are encrypted with different passwords. More precisely, given how the `ConsistentSettingsService` works, this check assures that a successful snapshot contains shards that are all encrypted with the current master's password for the repository. This works as intended and with no caveats.

What I wish to discuss is the case where passwords on some data nodes are different to the master's. In this case, only the shards allocated to data nodes which have the repository password identical to the repository password on the master node will be included in the snapshot. Here is an excerpt of the response to the caller:
```
{
  "snapshot" : {
    "snapshot" : "snapshot_4",
    "uuid" : "5pOG9Y9BQHGjxBXUYkgRCg",
    "version_id" : 8000099,
    "version" : "8.0.0",
    "indices" : [
      "twitter",
      "twitter2",
      "twitter3"
    ],
    "include_global_state" : true,
    "state" : "PARTIAL",
    "start_time" : "2020-01-15T12:56:42.078Z",
    "start_time_in_millis" : 1579093002078,
    "end_time" : "2020-01-15T12:56:42.687Z",
    "end_time_in_millis" : 1579093002687,
    "duration_in_millis" : 609,
    "failures" : [
      {
        "index" : "twitter3",
        "index_uuid" : "twitter3",
        "shard_id" : 0,
        "reason" : "org.elasticsearch.repositories.RepositoryException: [test_enc] Password mismatch for repository [test_enc]. The local node's value of the keystore secure setting [repository.encrypted.test_enc.password] is different from the master's\n\tat org.elasticsearch.repositories.encrypted.EncryptedRepository.snapshotShard(EncryptedRepository.java:101)\n\tat org.elasticsearch.snapshots.SnapshotShardsService.snapshot(SnapshotShardsService.java:341)\n\tat org.elasticsearch.snapshots.SnapshotShardsService.lambda$startNewShards$1(SnapshotShardsService.java:287)\n\tat org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:629)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)\n\tat java.base/java.lang.Thread.run(Thread.java:830)\n",
        "node_id" : "pN1b441YTV6fKug6AkF2Vg",
        "status" : "INTERNAL_SERVER_ERROR"
      }
    ],
    "shards" : {
      "total" : 5,
      "failed" : 1,
      "successful" : 4
    }
  }
}
```
And the relevant excerpt on the data node:
```
[2020-01-15T14:56:42,453][WARN ][o.e.c.s.ConsistentSettingsService] [ElasticMBP.local] the published hash [2MuYL08uwIKlzrCKoSEbqKsNK2m5V9vdMcd48nRj2D73L8JlPAWUsMpDNlpbgX7PHX+C256ifRVih9aB6AEIjA==] of the consistent secure setting [repository.encrypted.test_enc.password] differs from the locally computed one [0U7M0OWwH5gj3vJBGbpIFzAME6CBht1kqH6IxDRuTB9T/SThMQ9VSGcffALpl4hs3lGIo6aCBJwm1Y3q7qXPhQ==]
[2020-01-15T14:56:42,468][WARN ][o.e.s.SnapshotShardsService] [ElasticMBP.local] [[twitter3][0]][test_enc:snapshot_4/5pOG9Y9BQHGjxBXUYkgRCg] failed to snapshot shard
org.elasticsearch.repositories.RepositoryException: [test_enc] Password mismatch for repository [test_enc]. The local node's value of the keystore secure setting [repository.encrypted.test_enc.password] is different from the master's
        at org.elasticsearch.repositories.encrypted.EncryptedRepository.snapshotShard(EncryptedRepository.java:101) [repository-encrypted-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        at org.elasticsearch.snapshots.SnapshotShardsService.snapshot(SnapshotShardsService.java:341) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        at org.elasticsearch.snapshots.SnapshotShardsService.lambda$startNewShards$1(SnapshotShardsService.java:287) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:629) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
        at java.lang.Thread.run(Thread.java:830) [?:?]
```

Given the way the "Repository" fits into the snapshot/restore framework, I don't see a better way without significant code refactoring. For example, a better way might be to completely fail the snapshot. But this is difficult to achieve because the snapshot process fans-out to all nodes, with no pre-flight check calls, and because nodes cannot push to master that their keystore settings are different.
Another possibly better alternative is to fail the creating the repository in the case that passwords differ. But in this case, the consistency check will race with the actual publication of the keystore hashes; failing features at startup on the basis of keystore consistency is a core design limitation.

What do you think? Is this good enough given the substantial changes required by alternatives that fail faster in the snapshotting process? Do you have other ideas we should explore? Note that in this case repository verification fails, so the admin has a common way to learn that the encrypted repository is healthy.

Reply:
@albertzaharovits I personally think fail fast is not absolutely necessary here. With the incremental nature of snapshotting, it's not like the partial snapshot resulting from some broken nodes will be a complete waste of effort.
With the upcoming changes to the snapshot process that will allow parallel repository operations, I think I'll be able to provide you with better hooks for failing faster in the future if we want that, but for now it's just fine IMO.

Reply:
Thank you for the feedback Armin!

I will anyway try to make the password-consistency-check work on the master node and then call it on the `Repository#getRepositoryData` method, which is always invoked on the master node.

Reply:
:) posted 20s after https://github.com/elastic/elasticsearch/pull/50846#discussion_r367419852 => I like the idea!

Reply:
Hi @original-brownbear . I've been working to come up with a solution for how to check that all the nodes (master and data) have the same repository password, but ultimately failed.

Against the appearances, this is a fundamentally different problem than what the `ConsistentSettingsService` currently does. That's because the publishing of the cluster state, i.e. broadcasting from the master to the other nodes, which is what `ConsistentSettingsService` does, is exactly the opposite of the communication pattern in which nodes must inform the master of password hashes. For this chief reason, unifying them under one service class is unsuitable. Moreover, this communication pattern in which the master node agrees that at least all of the nodes that could take part in the snapshot have the same value, has to account for changing cluster topology, i.e. adding a new node, replacing the master, and possibly reloading the keystore; this all is complex work. Also, it is not enough only for the master node to verify it **before the snapshot starts**, ultimately the node doing the snapshot must also verify the password against the others in the cluster, not to rely on the master's previous check because otherwise this could be manifesting as a use-after-check type of error (although I couldn't give concrete example situations).

I was trying to prototype a way to hook-up repository verification (`client.admin().cluster().prepareVerifyRepository`) in the `Repository#getRepositoryData` call of the encrypted repository, when @tvernum raised a **very** good point that the case of a wrong password on some of the nodes is analogous to the case of wrong client credentials for the cloud-based repositories. I think that's exactly the case. In this situation the feature to validate that credentials are consistent among the participating nodes should be useful across the board, for all cloud repositories.

Given this, I propose we backtrack to the original proposal to only validate that the password on the data node snapshotting the shard is the same as the master's and accept the usability shortfall that the end result might be an incomplete snapshot (including an empty snapshot). In addition, we should raise a feature request issue for the support of credential consensus.

@original-brownbear would you agree to this course of action?

Reply:
If this sounds correct to you, that indeed this is a separate issue, do you consider that issue a blocker for encrypted snapshots feature? I would say no (although I'm obviously biased in this matter), because the analogous problem of cloud repository credentials has been extant for so long.